woocommerce/google-listings-and-ads
# GTag tracking for add_to_cart events

### Changes proposed in this Pull Request:

When we create pages using WooCommerce Gutenberg blocks we are unable to track the add_to_cart event since the page content is rendered dynamically. This PR enqueues some JavaScript assets which will hook into the appropriate events so it can track when products are added to the cart.

Since there are many add_to_cart buttons which can show up on archive/single product pages, including in the featured/related products sections, this PR will hook into all buttons on the page to do the tracking. This is less intrusive then injecting a code snippet for each add_to_cart button.

Note: For the Gutenberg blocks we are currently using an action name which is marked as [experimental](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/docs/blocks/feature-flags-and-experimental-interfaces.md#usages-of-experimental-prefix). I've started a discussion to get some feedback on this: p1651828112139989-slack-C02UBB1EPEF

Note: Automated tests haven't been added yet. Since it's being merged into a feature branch I'm going to add tests in a follow up PR.

### Detailed test instructions:

#### Test Gutenberg blocks
1. Use a site that is fully onboarded including connected to an Ads account.
2. Create a page using the All Products block 
3. View the page on the frontend and add a product to the cart (should be a simple product)
4. Check the network tab in the browser dev tools and confirm we are seeing the events getting sent to `https://googleads.g.doubleclick.net/pagead...`
5. On the products page jump to the next page (must have more products then fit on one page)
6. Click the add cart for one of the products dynamically loaded for another page
7. Confirm that the network tab show the event was tracked

#### Test regular add_to_cart buttons

1. Use a site that is fully onboarded including connected to an Ads account.
2. Go to a product archive page and test the add_to_cart buttons
3. Check the network tab in the browser dev tools and confirm we are seeing the events getting sent to `https://googleads.g.doubleclick.net/pagead...`
4. Confirm the same event works from a single product page
5. Confirm the same event works from related products on a single product page
6. Go to WooCommerce > Settings > Products and set the redirect to cart page option
7. Confirm events still get sent when we add_to_cart (might need to enable "preserve log" in console and network tab)
8. Go to WooCommerce > Settings > Products and disable the redirect to cart page option and disable "AJAX add to cart buttons on archives"
9. Confirm events still get sent when we add_to_cart

### Additional details:

The following points need to be finalised before this PR is ready:
- [x] Confirm usage of the experimental hook
- [x] Align parameters for the `add_to_cart` event to be the same as implemented in #1329 
- [x] Confirm whether we need to add the events `remove_from_cart` and `change_cart_quantity`
- [x] Check the method used for loading script assets on the frontend, and confirm this is the implementation we want to go with
- [x] The script gets loaded depending on `is_page` which means it will be loaded on each page, check if there is any way to know whether WC blocks have been used on the page
- [x] Confirm if we want to also include code for tracking regular cart events (not through blocks) to this script so we don't need to inject as much code on the page
- [ ] Add testing for any of the added code
- [ ] Add an e2e test for confirming the event is fired from a block (this will cover discovering any changes to the experimental hook)

### Changelog entry
* Add - Track add to cart events from all buttons including Gutenberg blocks.
